integrations/Matomo/util.js

```diff
@@ -0,0 +1,355 @@
+        shipping = properties.shipping;
+        discount = properties.discount;
+      }
+      if (!orderId) {
+        logger.error("User parameter order_id is required");
+        break;
+      }
+
+      /** grandTotal is a mandatory field
+       * If grandTotal is not found inside the properties, we are calculating it manually
+       */
+      if (!grandTotal) {
+        if (products && Array.isArray(products) && products.length > 0) {
+          grandTotal = 0;
+          // iterating through the products array and calculating grandTotal
+          for (let product = 0; product < products.length; product++) {
+            if (product.price && product.quantity) {
+              if (typeof price === "string") price = parseFloat(price);
+              if (typeof quantity === "string") quantity = parseFloat(quantity);
+              grandTotal += product.price * product.quantity;
+            }
+          }
+        }
+      }
+
+      /** subTotal is not a listed property in "Order Completed" event
+       * if user doesn't provide this property, then we are calculating it from grandTotal
+       * ref: https://matomo.org/faq/reports/analyse-ecommerce-reporting-to-improve-your-sales/#conversions-overview
+       */
+      if (!subTotal) {
+        subTotal = grandTotal;
+      }
+
+      // ref: https://matomo.org/faq/reports/analyse-ecommerce-reporting-to-improve-your-sales/#conversions-overview
+      if (shipping) {
+        if (typeof shipping === "string") shipping = parseFloat(shipping);
+        grandTotal += shipping;
+      }
+      if (tax) {
+        if (typeof tax === "string") tax = parseFloat(tax);
+        grandTotal += tax;
+      }
+
+      window._paq.push([
+        "trackEcommerceOrder",
+        orderId,
+        grandTotal,
+        subTotal,
+        tax,
+        shipping,
+        discount,
+      ]);
+      break;
+
+    case "CLEAR_ECOMMERCE_CART":
+      window._paq.push(["clearEcommerceCart"]);
+      break;
+
+    case "TRACK_ECOMMERCE_CART_UPDATE":
+      if (properties) grandTotal = properties.total || properties.revenue;
+      window._paq.push(["trackEcommerceCartUpdate", grandTotal]);
+      break;
+
+    default:
+      // Generic track Event
+      if (properties) {
+        category = properties.category;
+        action = properties.action;
+        value = properties.value;
+      }
+      if (!category) {
+        logger.error("User parameter category is required");
+        break;
+      }
+
+      if (!action) {
+        logger.error("User parameter action is required");
+        break;
+      }
+
+      name = message.event;
+      window._paq.push(["trackEvent", category, action, name, value]);
+      break;
+  }
+};
+
+/**
+ * Checks for custom dimensions in the payload
+ * If present, we make setCustomDimension() function call
+ * @param  {} message
+ */
+const checkCustomDimensions = (message) => {
+  const integrationsObj = getIntegrationsObj(message, "matomo");
```

Comment:
We actually can get the `integrations` object directly by `integrations[NAME]` where name is available in the `contstant.js` in the Matomo dir

Reply:
Done